File: tests/history.py
"""Senior High History browser regression against the built, offline index.html.

Run: python3 tests/history.py (Playwright + Chromium; see requirements.txt).
PW_CHROMIUM may point to another Chromium binary. Downloads use a temporary
folder, never the repository. No web server or AI service is needed.
"""
import os
from pathlib import Path
import tempfile
import xml.etree.ElementTree as ET
import zipfile

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parent.parent
URL = ROOT.joinpath("index.html").as_uri()
KEY_TITLE = "ANSWER KEYS — TEACHER'S COPY"
PERIODS = ["I", "II", "III", "IV", "V", "VI"]


def click(page, selector):
    # Settings can be inside a closed details panel; invoke the real handler.
    page.locator(selector).evaluate("el => el.click()")


def set_value(page, selector, value):
    page.locator(selector).evaluate("""(el, value) => {
        el.value = value;
        el.dispatchEvent(new Event('change', {bubbles: true}));
    }""", str(value))


def generate(page):
    click(page, "#gen")
    assert page.locator("#doc .page").count() > 0


def check_a4(page):
    sizes = page.locator("#doc .page").evaluate_all(
        "pages => pages.map(p => [p.offsetWidth, p.offsetHeight])"
    )
    assert sizes and all(size == [794, 1123] for size in sizes), sizes[:3]


def export_docx(page, directory, expected_name, expected_text, teacher):
    with page.expect_download() as event:
        click(page, "#docx")
    download = event.value
    assert download.suggested_filename == expected_name, download.suggested_filename
    path = Path(directory, expected_name)
    download.save_as(path)
    with zipfile.ZipFile(path) as docx:
        assert docx.testzip() is None
        xml = docx.read("word/document.xml")
        document = ET.fromstring(xml)
        ns = {"w": "http://schemas.openxmlformats.org/wordprocessingml/2006/main"}
        text = " ".join(document.itertext())
        assert expected_text in text, expected_text
        assert "HISTORY" in text.upper()
        if "Workbook" in expected_name:
            assert (KEY_TITLE in text) == teacher, "teacher-only keys in Word"
            assert any(name.startswith("word/media/") for name in docx.namelist()), "offline cover image"
        for size in document.findall(".//w:pgSz", ns):
            assert size.get("{" + ns["w"] + "}w") == "11906"
            assert size.get("{" + ns["w"] + "}h") == "16838"
    return path.stat().st_size


with sync_playwright() as playwright, tempfile.TemporaryDirectory(prefix="lib-history-") as directory:
    launch = {"args": ["--no-sandbox", "--disable-dev-shm-usage"]}
    if os.environ.get("PW_CHROMIUM"):
        launch["executable_path"] = os.environ["PW_CHROMIUM"]
    browser = playwright.chromium.launch(**launch)
    context = browser.new_context(viewport={"width": 1500, "height": 1000}, accept_downloads=True)
    # The deliverable must work with the network disconnected.
    context.set_offline(True)
    page = context.new_page()
    errors = []
    page.on("pageerror", lambda error: errors.append(str(error)))
    page.goto(URL)
    page.wait_for_selector(".subtab[data-s='hi']")
    assert page.locator(".subtab[data-s='whis']").count() == 0
    assert page.locator(".subtab[data-s='hi']").inner_text().strip() == "History"
    assert "History" in page.title()
    click(page, ".subtab[data-s='hi']")
    assert page.locator("#grade option").evaluate_all("els => els.map(el => el.value)") == ["10", "11", "12"]
    assert "Senior High" in page.locator("#jhNote").inner_text()
    assert "History" in page.locator("#jhNote").inner_text()
    assert page.locator("#sheets .sh:checked").count() == 11
    assert "Social Studies" not in page.locator("#sheets").inner_text()
    assert page.locator(".subtab[data-s='hi'] use").get_attribute("href") == "#i-sub-his"
    assert page.evaluate("""() => document.querySelector('#doc .cv-bg').style.backgroundImage
        .includes(window.SUBJECT_COVER_ART.ss.url)"""), "world-history artwork available offline"
    assert page.evaluate("getComputedStyle(document.body).getPropertyValue('--bleu').trim()") == "#8a5a00"

    for grade in [10, 11, 12]:
        set_value(page, "#grade", grade)
        assert page.locator("#periods .pk").evaluate_all("els => els.map(el => el.value)") == PERIODS
        expected_titles = page.evaluate("grade => HI_CURRICULUM.filter(u => u.grade === grade).map(u => u.title)", grade)
        for mode in ["student", "teacher"]:
            click(page, f"#session .sess[data-m='{mode}']")
            generate(page)
            text = page.locator("#doc").inner_text()
            assert f"HISTORY · SENIOR HIGH GRADE {grade}" in text
            assert (KEY_TITLE in text) == (mode == "teacher")
            assert "Study Notes" in text
            assert "SEMESTER ONE EXAMINATION" in text and "SEMESTER TWO EXAMINATION" in text
            for title in expected_titles:
                assert title in text, title
            tests = page.locator("#doc h2").evaluate_all(
                "els => els.filter(el => el.textContent.startsWith('PERIOD TEST')).length"
            )
            assert tests == 6
            assert "Social Studies" not in text and "WASSCE History" not in text
            assert page.evaluate("PACK_CUR_SUBJECT") == "History"
            assert page.evaluate("PACK_CUR_GRADE") == grade
            check_a4(page)
            suffix = "Teacher_Copy" if mode == "teacher" else "Student"
            size = export_docx(page, directory, f"History_Grade{grade}_Workbook_{suffix}.docx",
                               expected_titles[0], mode == "teacher")
            print(f"OK: History Grade {grade}, {mode}, six periods, A4, Word ({size} bytes)")

    # Period selection and the study-notes toggle operate on real History data.
    page.locator("#periods .pk").evaluate_all("els => els.forEach(el => el.checked = el.value === 'II')")
    generate(page)
    assert "1 unit(s)" in page.locator("#meta").inner_text()
    assert "Foreign Policy of Liberia" in page.locator("#doc").inner_text()
    assert "Seven Selected Liberian Presidents" not in page.locator("#doc").inner_text()
    click(page, "#notes")
    assert "Study Notes" not in page.locator("#doc").inner_text()
    click(page, "#notes")
    assert "Study Notes" in page.locator("#doc").inner_text()

    # Both teacher planning formats and standard 3/4-week durations.
    click(page, "#dtype .sess[data-d='lp']")
    for grade in [10, 11, 12]:
        set_value(page, "#grade", grade)
        page.locator("#periods .pk").evaluate_all("els => els.forEach(el => el.checked = el.value === 'I')")
        for plan_type in ["daily", "weekly"]:
            click(page, f"#lpPlanTabs .sess[data-pt='{plan_type}']")
            for weeks in [3, 4]:
                click(page, f"#lpWeekPresets .lp-w[data-w='{weeks}']")
                set_value(page, "#lpMin", 45)
                generate(page)
                text = page.locator("#doc").inner_text()
                assert "TEACHER’S LESSON PLAN" in text
                assert "Subject: History" in text
                assert "Weekly Plan Adjustment" in text
                assert f"{weeks} weeks/unit" in page.locator("#meta").inner_text()
                assert "Examination-style practice" not in text
                if plan_type == "weekly":
                    assert f"Week {weeks} of {weeks}:" in text
                check_a4(page)
        export_docx(page, directory, f"History_Grade{grade}_Weekly_Plan_Teacher_Copy.docx",
                    "Weekly Plan Adjustment", True)
        print(f"OK: History Grade {grade}, daily/weekly lesson plans, 3/4 weeks and Word")

    # Switching tracks must not replace, duplicate or re-grade WASSCE History.
    click(page, "#dtype .sess[data-d='pack']")
    click(page, "#tracks .track[data-t='wa']")
    assert page.locator(".subtab[data-s='hi']").count() == 0
    click(page, ".subtab[data-s='whis']")
    assert page.locator("#grade option").evaluate_all("els => els.map(el => el.value)") == ["12"]
    assert page.locator("#periods .pk").count() == 13
    assert page.evaluate("PACK_CUR_SUBJECT") == "WASSCE History"
    click(page, "#tracks .track[data-t='curr']")
    click(page, ".subtab[data-s='hi']")
    assert page.locator("#grade option").count() == 3
    assert page.locator("#periods .pk").count() == 6

    # Actual print output and filename use History, not its shared engine.
    page.locator("#periods .pk").evaluate_all("els => els.forEach(el => el.checked = el.value === 'I')")
    generate(page)
    page.evaluate("window.print = () => { window.__printedTitle = document.title; }")
    click(page, "#print")
    assert page.evaluate("window.__printedTitle").startswith("History_Grade")
    page.evaluate("window.dispatchEvent(new Event('afterprint'))")
    pdf = page.pdf(format="A4", print_background=True, prefer_css_page_size=True)
    assert pdf.startswith(b"%PDF") and len(pdf) > 10000
    print(f"OK: separate WASSCE History, offline cover and PDF ({len(pdf)} bytes)")

    # Mobile preview retains fixed A4 geometry and fits the viewport.
    page.set_viewport_size({"width": 375, "height": 812})
    page.wait_for_function("parseFloat(getComputedStyle(document.documentElement).getPropertyValue('--ps')) < 1")
    check_a4(page)
    assert page.locator("#doc .page").first.evaluate("el => el.getBoundingClientRect().width") <= 377
    assert page.evaluate("Math.max(document.body.scrollWidth, document.documentElement.scrollWidth)") <= 377
    assert not errors, errors
    browser.close()
    print("OK: History mobile layout and no browser errors; all checks passed.")

File: tests/history.js
#!/usr/bin/env node
/* Senior High History regression (no dependencies):
   node tests/history.js
   Checks the actual registry, source coverage, every default worksheet,
   student/teacher assessments, deterministic seeds and daily/weekly plans.
   Browser controls, printing and Word bytes are checked by history.py. */
"use strict";
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");
const vm = require("node:vm");
const root = path.resolve(__dirname, "..");
const read = file => fs.readFileSync(path.join(root, file), "utf8");
const app = read("app.js");
const html = read("index.html");
const context = {
  console,
  PACK_COVER: (opts, cover) => [
    { k: "p", t: cover.title }, { k: "p", t: cover.line }
  ]
};
context.window = context;
vm.createContext(context);
function load(file) {
  vm.runInContext(read(file), context, { filename: file });
}
function section(start, end) {
  const a = app.indexOf(start), b = app.indexOf(end, a);
  assert(a >= 0 && b > a, "app source section exists: " + start);
  vm.runInContext(app.slice(a, b), context);
}
load("data-hi.js");
load("gen-ss.js");
load("lesson.js");
section("var SUBJECTS =", "/* ---------------- education bands");
section("/* ---------------- Study Notes & Explanations", "/* ---------------- session mode");

const subject = context.SUBJECTS.hi;
assert(subject, "History is registered in the regular curriculum");
assert.equal(subject.label, "History");
assert(!subject.wa, "History must not be hidden behind the WASSCE track");
assert.equal(subject.curriculum(), context.HI_CURRICULUM);
assert.equal(subject.engine(), context.GEN_SS);
assert.equal(subject.coverArt, "ss", "Reuse the bundled world-history cover");
assert(read("icons.svg.html").includes('id="i-' + subject.icon + '"'));
for (const file of ["data-hi.js", "gen-ss.js", "lesson.js", "app.js"]) {
  assert(html.includes(read(file).trim()), file + " is included in the rebuilt offline deliverable");
}
assert(read("build.sh").includes("data-hi.js"), "future builds retain History");

const units = subject.curriculum();
const snapshot = JSON.stringify(units);
const periods = ["I", "II", "III", "IV", "V", "VI"];
const expectedPages = [
  [2,3], [4,5], [6,7], [8,9,10,11], [12], [13],
  [14], [15,16], [17,18], [19,20], [21,22], [23,24],
  [25,26], [27,28], [29], [30,31], [32], [33,34]
];
assert.equal(units.length, 18);
assert.equal(new Set(units.map(u => u.grade + "/" + u.period)).size, 18);

units.forEach((unit, i) => {
  const label = "Grade " + unit.grade + " Period " + unit.period;
  assert.equal(unit.grade, 10 + Math.floor(i / 6), label);
  assert.equal(unit.period, periods[i % 6], label);
  assert.equal(unit.sem, i % 6 < 3 ? "One" : "Two", label);
  assert.equal(unit.source.file, "uploads/History 10-12.pdf");
  assert.deepEqual(Array.from(unit.source.pages), expectedPages[i], label + " source mapping");
  assert(fs.existsSync(path.join(root, unit.source.file)));
  assert(unit.objectives.length >= 4 && unit.outcomes.length, label + " objectives");
  assert(unit.study.length >= 8, label + " substantive study blocks");
  assert(unit.study.some(b => b.k === "p" && b.t.includes("**")), label + " key terms in notes");
  assert(unit.terms.length >= 5 && unit.facts.length >= 5 && unit.tf.length >= 5,
    label + " has enough items for full 40-mark period tests");
  assert.equal(new Set(unit.terms.map(v => v.t)).size, unit.terms.length, label + " unique terms");
  for (const term of unit.terms) {
    assert(term.t && term.d && term.x, label + " complete vocabulary");
    assert(term.x.toLowerCase().includes(term.t.toLowerCase()), label + " cloze-ready examples");
  }
  for (const fact of unit.facts.concat(unit.apply, unit.casestudy.questions)) {
    assert(fact.q && fact.a, label + " answer-bearing questions");
  }
  for (const item of unit.tf) {
    assert(["True", "False"].includes(item.a) && item.s && item.why, label + " explained TF key");
  }
  assert(unit.tf.some(v => v.a === "False") && unit.tf.some(v => v.a === "True"));
  assert(unit.mapwork.items.length >= 5 && unit.sort.groups.length >= 2);
  assert(unit.project.steps.length && unit.project.criteria.length);
  assert(unit.activities.length && unit.materials.length && unit.assessment.length);
  for (const id of subject.defaults) {
    const sheet = subject.engine().SHEETS[id];
    assert(sheet, "registered worksheet " + id);
    const result = sheet.fn(unit, 8, () => 0.37);
    assert(result && result.blocks.length, label + " produces " + id);
    if (id !== "terms") assert(result.key.length, label + " has " + id + " marking guidance");
    assert(!JSON.stringify(result).includes("undefined"), label + " complete worksheet output");
  }
});

function options(grade, keys, seed) {
  return {
    grade, keys, seed, perEx: 8, tests: true, exam: true,
    curriculum: units, sheets: subject.defaults,
    subjectId: "hi", subjectName: "HISTORY", subjectLine: "History", bandName: "Senior High",
    teacher: keys, lpMin: 45, lpDays: 5, lpAdjMode: "standard"
  };
}
for (const grade of [10, 11, 12]) {
  assert.equal(subject.file(grade), "History_Grade" + grade + "_Workbook.docx");
  for (const keys of [false, true]) {
    for (const seed of [1, 42, 9999]) {
      const opts = options(grade, keys, seed);
      const pack = subject.engine().buildPack(opts);
      const text = JSON.stringify(pack.blocks);
      assert.equal(pack.topics.length, 6);
      assert.equal(pack.blocks.filter(b => b.k === "h2" && b.t.startsWith("PERIOD TEST")).length, 6);
      assert.equal(pack.blocks.filter(b => b.k === "h2" && b.t.startsWith("SEMESTER")).length, 2);
      assert.equal(text.includes("ANSWER KEYS — TEACHER'S COPY"), keys);
      assert(text.includes("HISTORY — GRADE " + grade));
      assert(!text.includes("SOCIAL STUDIES") && !text.includes("Social Studies"));
      assert.equal(JSON.stringify(subject.engine().buildPack(opts).blocks), text, "deterministic seed");
    }
  }
  for (const unit of units.filter(u => u.grade === grade)) {
    const opts = { ...options(grade, true, 42), topics: [unit.period] };
    assert.equal(subject.engine().buildPack(opts).topics.length, 1, "individual period selection");
    for (const lpPlanType of ["daily", "weekly"]) {
      for (const lpWeeks of [3, 4]) {
        const plan = context.LESSON_PLAN.build({ ...opts, lpPlanType, lpWeeks });
        const text = JSON.stringify(plan.blocks);
        assert.equal(plan.topics.length, 1);
        assert(text.includes(unit.title) && text.includes("Subject: History"));
        assert(text.includes("Weekly Plan Adjustment") && text.includes("Assignment"));
        assert(!text.includes("Examination-style practice"), "History-specific teaching methods");
        assert(!text.includes("Outline map of Liberia"), "maps should fit this unit's region");
        if (lpPlanType === "weekly") assert(text.includes("Week " + lpWeeks + " of " + lpWeeks));
        else {
          const minutes = plan.blocks.filter(b => ["h2", "h3"].includes(b.k) && /\(\d+ min\)/.test(b.t))
            .map(b => +b.t.match(/\((\d+) min\)/)[1]);
          assert.equal(minutes.length, 4);
          assert.equal(minutes.reduce((a, b) => a + b, 0), 45);
        }
      }
    }
  }
}
assert.equal(JSON.stringify(units), snapshot, "generation must not mutate the curriculum");

// WASSCE History remains a separate, unchanged Grade 12 examination curriculum.
load("data-wa.js");
load("data-wa-his.js");
const wassce = context.SUBJECTS.whis;
assert.equal(wassce.wa, "his");
assert.equal(wassce.curriculum().length, 13);
assert(wassce.curriculum().every(unit => unit.grade === 12));
assert.notEqual(wassce.curriculum(), units);
console.log("OK: Senior High History — 18 units, 11 default worksheets, student/teacher packs, 72 lesson-plan variants and separate WASSCE coverage.");
